Source/WarpX.H

```diff
@@ -951,6 +951,44 @@ private:
     // Singleton is used when the code is run from python
     static WarpX* m_instance;
 
+    /**
+     * \brief
+     * Signal handling
+     *
+     * Rank 0 will accept signals and asynchronously broadcast the
+     * configured resposne; other processes will ignore them and
+     * follow the lead of rank 0 to avoid potential for deadlocks or
+     * timestep-skewed response.
+     *
+     * Variables and functions are static rather than per-instance
+     * because signal handlers are configured at the process level.
+     */
+    //! Set up signal handlers based on input configuration
+    void InitSignalHandling();
+    //! Check and clear signal flags and asynchronously broadcast them from process 0
+    static void CheckSignals();
+    //! Signal handler to set flags on process 0 (other processes ignore configured signals)
+    static void SignalSetFlag(int signal_number);
+    //! Complete the asynchronous broadcast of signal flags, and initiate a checkpoint if requested
+    void HandleSignals();
+    //! The range of signal values to accept
+    static constexpr int NUM_SIGNALS = 32;
+    //! On process 0, whether a given signal has been received since the last check
+    static std::atomic<bool> signal_received_flags[NUM_SIGNALS];
+    //! Whether configuration requests the code to break out of the timestep loop on a particular signal
+    static bool signal_conf_requests_break[NUM_SIGNALS];
+    //! Whether configuration requests the code to checkpoint at this timestep on a particular signal
+    static bool signal_conf_requests_checkpoint[NUM_SIGNALS];
+    //! Boolean flags transmitted between CheckSignals() and
+    //! HandleSignals() to indicate actions requested by signals
+    static bool signal_actions_requested[2];
+    //! Labels for indexed positions in signal_actions_requests
+    enum signal_action_requested_labels { SIGNAL_REQUESTS_BREAK = 0, SIGNAL_REQUESTS_CHECKPOINT = 1 };
```

Comment:
I could stick a maximum in the `enum`, and then use that in place of the literal `2` array size.